src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php

```diff
@@ -3,115 +3,161 @@
 use EventSauce\EventSourcing\UnableToPersistMessages;
 use EventSauce\EventSourcing\UnableToRetrieveMessages;
+use EventSauce\MessageRepository\DoctrineMessageRepository\TableSchema\DefaultTableSchema;
+use EventSauce\MessageRepository\DoctrineMessageRepository\UuidEncoder\BinaryUuidEncoder;
 use Generator;
 use Ramsey\Uuid\Uuid;
 use Throwable;
 
+use function array_keys;
+use function array_map;
+use function array_merge;
 use function count;
+use function implode;
+use function json_decode;
+use function json_encode;
+use function sprintf;
 
 class DoctrineUuidV4MessageRepository implements MessageRepository
 {
     private int $jsonEncodeOptions = 0;
+    private TableSchema $tableSchema;
+    private UuidEncoder $uuidEncoder;
 
     public function __construct(
         private Connection $connection,
         private string $tableName,
         private MessageSerializer $serializer,
-    ) {}
+        ?TableSchema $tableSchema = null,
+        ?UuidEncoder $uuidEncoder = null,
+    ) {
+        $this->tableSchema = $tableSchema ?? new DefaultTableSchema();
+        $this->uuidEncoder = $uuidEncoder ?? new BinaryUuidEncoder();
+    }
 
     public function persist(Message ...$messages): void
     {
         if (count($messages) === 0) {
             return;
         }
 
-        $sql = "INSERT INTO {$this->tableName} (event_id, aggregate_root_id, version, payload) VALUES ";
-        $values = [];
-        $parameters = [];
+        $insertColumns = [
+            $this->tableSchema->eventIdColumn(),
+            $this->tableSchema->aggregateRootIdColumn(),
+            $this->tableSchema->versionColumn(),
+            $this->tableSchema->payloadColumn(),
+            ...array_keys($this->tableSchema->additionalColumns()),
+        ];
+
+        $insertValues = [];
+        $insertParameters = [];
 
         foreach ($messages as $index => $message) {
             $payload = $this->serializer->serializeMessage($message);
-            $eventIdColumn = 'event_id_' . $index;
-            $aggregateRootIdColumn = 'aggregate_root_id_' . $index;
-            $versionColumn = 'version_' . $index;
-            $payloadColumn = 'payload_' . $index;
-            $values[] = "(:{$eventIdColumn}, :{$aggregateRootIdColumn}, :{$versionColumn}, :{$payloadColumn})";
-            $parameters[$versionColumn] = $payload['headers'][Header::AGGREGATE_ROOT_VERSION] ?? 0;
-            $payload['headers'][Header::EVENT_ID] = $payload['headers'][Header::EVENT_ID] ?? Uuid::uuid4()->toString();
-            $parameters[$eventIdColumn] = $this->uuidToBinary($payload['headers'][Header::EVENT_ID]);
-            $parameters[$payloadColumn] = json_encode($payload, $this->jsonEncodeOptions);
-            $parameters[$aggregateRootIdColumn] = $this->uuidToBinary($payload['headers'][Header::AGGREGATE_ROOT_ID] ?? '');
+            $payload['headers'][Header::EVENT_ID] ??= Uuid::uuid4()->toString();
+
+            $messageParameters = [
```

Comment:
We have to use a separate array here, rather than updating `$insertParameters`, because there is an unknown number of columns to do `TableSchema::additionalColumns()` return having a variable length.